lib/definition/types/each.rb

```diff
@@ -13,64 +13,50 @@ def initialize(name, definition:)
         super(name)
       end
 
-      def conform(value)
-        Conformer.new(self).conform(value)
-      end
+      def conform(values)
+        return non_array_error(values) unless values.is_a?(Array)
 
-      def error_renderer
-        ErrorRenderers::Leaf
-      end
+        errors = false
 
-      class Conformer
-        def initialize(definition)
-          self.definition = definition
+        results = values.map do |value|
```

Comment:
good question, i think we could do so maybe :thinking: I'll need to have a closer look. The readme describes it as follows:

> This definition will only conform if all definitions conform. The definitions will be processed from left to right and the output of the previous will be the input of the next. Processing of the And-Definition stops as soon as one definition does not conform.

Reply:
hm yeah at least that is how it currently works :thinking: The error case is not so much the bottleneck though from what i've found. In error cases definition is already faster then dry-struct. The main performance issue is behind the all-is-valid case

Reply:
Just an idea to try.
Instead of using `map` and building up an array with entries we probably don't need, we could also have two separte stacks for valid and invalid results like this:
```
results = []
faulty_results = []

values.each do |value|
  result = item_definition.conform(value)
          
  if result.passed?
    results << result
  else
    errors = true
    faulty_results << result
   end
end

return ConformResult.new(results.map(&:value)) unless errors

ConformResult.new(values, errors: [ConformError.new(self,
                                                            "Not all items conform with '#{name}'",
                                                            sub_errors: convert_errors(faulty_results))])

...

def convert_errors(faulty_results)
  errors = []
  faulty_results.each_with_index do |faulty_result, index|
    errors << KeyConformError.new(
      self,
      "Item #{faulty_result.value.inspect} did not conform to #{name}",
       key:        index,
       sub_errors: faulty_result.error_tree
       )
       end
      errors
end
```
`Map` uses `push` under the hood, which is a bit slower than `<<`.
Since we don't need all entries if there is an error, we don't need to have all individual results stored in the array.
By having a separate `faulty_results` stack, we can get rid of a check within `convert_errors`, too.

This version constantly performed a bit better on my machine.
Maybe give it a try and see if it's worth it.

Reply:
Feel free to ignore aboves suggestion if it's mostly about the all-is-valid case

Reply:
just gave this a try with a slight adpatation, i still need all results for the convert_errors function, as this one needs to extract the correct index of the array elements for the error message.

with that i don't see a significant change in execution time. the benchmark results vary more from run to run than the difference it makes

Reply:
also with this the results are within error, seems even a bit slower:
```ruby
def conform(values)
        return non_array_error(values) unless values.is_a?(Array)

        errors = []
        result_values = []

        values.each_with_index do |value, index|
          result = item_definition.conform(value)
          if result.passed?
            result_values << result.value
          else
            errors << KeyConformError.new(
            self,
            "Item #{result.value.inspect} did not conform to #{name}",
            key:        index,
            sub_errors: result.error_tree
          )
          end
        end

        return ConformResult.new(result_values) if errors.empty?

        ConformResult.new(values, errors: [ConformError.new(self,
                                                            "Not all items conform with '#{name}'",
                                                            sub_errors: errors)])
      end
```